File: tests/fixtures/lang_samples/csharp/PerfIoInLoop.cs
// Fixture for the C# performance dialect (io_in_loop / string_concat /
// blocking_sync_in_async). Every POSITIVE is a genuine per-iteration I/O
// boundary or a sync-over-async block; every NEGATIVE is a shape the dialect
// must NOT flag. Hand-counted expectations live in test_perf_csharp.py. NOTE:
// this file imports EF Core, so file-level db evidence is present; the
// in-memory-LINQ false-positive gating is tested inline instead.
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

class PerfIoInLoop
{
    // --- POSITIVES ---------------------------------------------------------

    async Task EfAsyncN1(DbContext ctx, List<int> ids)
    {
        foreach (var id in ids)
        {
            await ctx.Users.ToListAsync();   // POSITIVE: EF *Async (unambiguous db)
        }
    }

    async Task HttpInLoop(HttpClient c, List<string> urls)
    {
        foreach (var u in urls)
        {
            await c.GetAsync(u);             // POSITIVE: network
        }
    }

    async Task SaveInLoop(DbContext ctx, List<int> ids)
    {
        foreach (var id in ids)
        {
            await ctx.SaveChangesAsync();    // POSITIVE: EF write
        }
    }

    void FileReadInLoop(List<string> paths)
    {
        foreach (var p in paths)
        {
            File.ReadAllText(p);             // POSITIVE: filesystem
        }
    }

    void SyncLinqInLoop(DbContext ctx, List<int> ids)
    {
        foreach (var id in ids)
        {
            ctx.Users.ToList();              // POSITIVE: sync LINQ, db import present
        }
    }

    string StringConcatInLoop(List<string> rows)
    {
        string s = "";
        foreach (var r in rows)
        {
            s += "line";                     // POSITIVE: quadratic string build
        }
        return s;
    }

    async Task SyncOverAsync(Task<int> task)
    {
        var r = task.Result;                 // POSITIVE: .Result blocks in async
        task.Wait();                         // POSITIVE: .Wait() blocks in async
    }

    // --- NEGATIVES ---------------------------------------------------------

    void PureCallInLoop(List<int> items)
    {
        int total = 0;
        foreach (var x in items)
        {
            total += x;                      // NEGATIVE: numeric +=, no I/O
        }
    }

    async Task AwaitedAsyncResult(Task<int> task)
    {
        var r = await task;                  // NEGATIVE: awaited, not .Result/.Wait
    }
}

File: tests/fixtures/lang_samples/rust/perf_io_in_loop.rs
// Fixture for the Rust performance dialect (io_in_loop / regex_compile_in_loop /
// resource_construction_in_loop / blocking_sync_in_async). Every POSITIVE is a
// genuine per-iteration cost; every NEGATIVE is a shape the dialect must NOT
// flag. Hand-counted expectations live in test_perf_dialects.py.
//
// NOTE: Rust `String` building (`push_str` / `+=`) is amortized O(1), so
// `string_concat_in_loop` is intentionally not a Rust marker — the push_str
// loop below is a NEGATIVE by design.

use std::fs;

// --- POSITIVES -------------------------------------------------------------

async fn sqlx_fetch_in_loop(pool: &PgPool, ids: Vec<i32>) {
    for id in ids {
        // POSITIVE: a sqlx executor verb (db round-trip) per item, awaited
        // (also a serial_await_in_loop co-signal).
        let _ = sqlx::query("SELECT 1").fetch_all(pool).await;
    }
}

fn fs_read_in_loop(paths: Vec<String>) {
    for p in paths {
        let _ = std::fs::read(&p); // POSITIVE: filesystem per item
    }
}

fn fs_read_to_string_in_loop(paths: Vec<String>) {
    for p in paths {
        let _ = fs::read_to_string(&p); // POSITIVE: filesystem (short `fs::` path)
    }
}

async fn reqwest_get_in_loop(urls: Vec<String>) {
    for u in urls {
        // POSITIVE: network per item, awaited (serial_await co-signal).
        let _ = reqwest::get(&u).await;
    }
}

fn regex_new_in_loop(items: Vec<String>) {
    for x in items {
        let re = Regex::new("^[a-z]+$"); // POSITIVE: static literal pattern, hoistable
        let dyn_re = Regex::new(&x); // NEGATIVE: dynamic arg cannot be hoisted
        let _ = (re, dyn_re);
    }
}

async fn pgpool_connect_in_loop(urls: Vec<String>) {
    for u in urls {
        // POSITIVE: a fresh pooled connection per item (resource construction).
        let _ = PgPool::connect(&u).await;
    }
}

async fn thread_sleep_in_async(d: std::time::Duration) {
    std::thread::sleep(d); // POSITIVE: blocking the executor inside an async fn
}

// --- NEGATIVES -------------------------------------------------------------

fn fs_read_constant_loop() {
    for _ in 0..10 {
        let _ = std::fs::read("x"); // NEGATIVE: bounded integer-literal range
    }
}

fn fs_read_outside_loop(p: String) {
    let _ = std::fs::read(&p); // NEGATIVE: runs once, not in a loop
}

fn push_str_in_loop(items: Vec<String>) -> String {
    let mut s = String::new();
    for x in items {
        s.push_str(&x); // NEGATIVE: amortized O(1) — not a Rust perf smell
    }
    s
}
